src/main/java/de/tum/in/www1/artemis/web/rest/GenericSubmissionResource.java

```diff
@@ -0,0 +1,241 @@
+ */
+public abstract class GenericSubmissionResource<T extends Submission> {
+
+    @Value("${jhipster.clientApp.name}")
+    protected String applicationName;
+
+    private static final String ENTITY_NAME = "genericSubmission";
+
+    protected final CourseService courseService;
+
+    protected final AuthorizationCheckService authCheckService;
+
+    protected final ExerciseService exerciseService;
+
+    protected final UserService userService;
+
+    protected final ParticipationService participationService;
+
+    public GenericSubmissionResource(CourseService courseService, AuthorizationCheckService authCheckService, UserService userService, ExerciseService exerciseService,
+            ParticipationService participationService) {
+        this.userService = userService;
+        this.exerciseService = exerciseService;
+        this.courseService = courseService;
+        this.authCheckService = authCheckService;
+        this.participationService = participationService;
+    }
+
+    /**
+     * Check if exercise is valid, has course and the user (student) can access it
+     * @param exercise that we want to check
+     * @return either null if exercise is valid or one of the error responses if it is not valid
+     */
+    final ResponseEntity<T> checkExerciseValidityForStudent(Exercise exercise) {
+        if (exercise == null) {
+            return ResponseEntity.badRequest()
+                    .headers(HeaderUtil.createFailureAlert(applicationName, true, "submission", "exerciseNotFound", "No exercise was found for the given ID.")).body(null);
+        }
+
+        // fetch course from database to make sure client didn't change groups
+        final Course course = courseService.findOne(exercise.getCourse().getId());
+        if (course == null) {
+            return ResponseEntity.badRequest()
+                    .headers(
+                            HeaderUtil.createFailureAlert(applicationName, true, ENTITY_NAME, "courseNotFound", "The course belonging to this file upload exercise does not exist"))
+                    .body(null);
+        }
+        if (!authCheckService.isAtLeastStudentInCourse(course, userService.getUserWithGroupsAndAuthorities())) {
+            return ResponseEntity.status(HttpStatus.FORBIDDEN).build();
+        }
+
+        return null;
+    }
+
+    /**
+     * Check if exercise is valid, the user (tutor) can access it and the user didn't reach the limit of submission locks
+     * @param exercise that we want to check
+     * @param exerciseType type of the exercise
+     * @param submissionService concrete submission service that is used to check lock limit
+     * @return either null if exercise is valid or one of the error responses if it is not valid
+     */
+    final <E extends Exercise> ResponseEntity<T> checkExerciseValidityForTutor(Exercise exercise, Class<E> exerciseType, SubmissionService<T> submissionService) {
```

Comment:
Will there be a refactoring of the exercise services? This kind of goes into the same direction as my comment regarding participations. With generic exercise services, we wouldn't have to use class types as arguments. E.g. in this case, we wouldn't have to check if an exercise is an instance of the actual desired exercise subclass since a generic exercise service/repository would ensure that we can only fetch the specified type in the first place.

Reply:
Why do you have the `SubmissionService` as a parameter here? This should get injected in the constructor of the resource i.m.o.

Reply:
Yes the refactoring for exercise service is also planned. I believe we can do this in other PR as well and then simplify how it is used here.

Reply:
Yes you are right about the `submissionService`,  but since I already want to use it in Programming Exercise and ProgrammingSubmisionService doesn't implement `SubmissionService` I had to leave it out for now